pkg/controller/job/job_controller.go

```diff
@@ -810,23 +810,33 @@ func (jm *Controller) syncJob(ctx context.Context, key string) (rErr error) {
 	if len(ns) == 0 || len(name) == 0 {
 		return fmt.Errorf("invalid job key %q: either namespace or name is missing", key)
 	}
+	clean := func() error {
+		jm.expectations.DeleteExpectations(logger, key)
+		jm.finalizerExpectations.deleteExpectations(logger, key)
+
+		err := jm.podBackoffStore.removeBackoffRecord(key)
+		if err != nil {
+			// re-syncing here as the record has to be removed for finished/deleted jobs
+			return fmt.Errorf("error removing backoff record %w", err)
+		}
+		return nil
+	}
 	sharedJob, err := jm.jobLister.Jobs(ns).Get(name)
 	if err != nil {
 		if apierrors.IsNotFound(err) {
-			logger.V(4).Info("Job has been deleted", "key", key)
-			jm.expectations.DeleteExpectations(logger, key)
-			jm.finalizerExpectations.deleteExpectations(logger, key)
-
-			err := jm.podBackoffStore.removeBackoffRecord(key)
-			if err != nil {
-				// re-syncing here as the record has to be removed for finished/deleted jobs
-				return fmt.Errorf("error removing backoff record %w", err)
-			}
-			return nil
+			logger.V(0).Info("Job has been deleted", "key", key)
+			return clean()
 		}
 		return err
 	}
 
+	if newlyCreated(sharedJob) {
```

Comment:
Add a comment explaining why we do so. 

```
Make sure the local in-memory state is cleaned when handling a newly created Job.
This is needed in case we still have some in-memory state left by the previous Job
with the same key (name & namespace). We use selected status fields to determine
the Job is newly created.
```